src/evm/protocol/uniswap_v3/state.rs

```diff
@@ -221,6 +227,104 @@ impl UniswapV3State {
             sqrt_price_next
         }
     }
+
+    /// Converts a target price to sqrt_price_x96 format with fee adjustment
+    ///
+    /// # Arguments
+    /// * `token_in` - The token being sold
+    /// * `token_out` - The token being bought
+    /// * `target_price` - The target price as token_out/token_in (tycho convention)
+    ///
+    /// # Returns
+    /// The sqrt price limit in Q96 format
+    fn get_sqrt_price_limit(
+        &self,
+        token_in: &Bytes,
+        token_out: &Bytes,
+        target_price: &Price,
+    ) -> Result<U256, SimulationError> {
+        let zero_for_one = token_in < token_out;
+
+        // Convert target pool price (token_out/token_in) to swap price (token_in/token_out)
+        // by flipping numerator and denominator
+        let swap_price_numerator = biguint_to_u256(&target_price.denominator);
+        let swap_price_denominator = biguint_to_u256(&target_price.numerator);
+
+        // Apply fee to the sell price (numerator) using integer division to match APEX behavior
+        // For Uniswap V3: adjusted_price = price * (1 - fee)
+        let fee_tier = U256::from(self.fee as u32);
+        let sell_price_after_fee =
+            safe_div_u256(swap_price_numerator * (FEE_PRECISION - fee_tier), FEE_PRECISION)?;
+        let buy_price = swap_price_denominator;
+
+        // Determine which price goes to which parameter based on swap direction
+        // For Uniswap V3, sqrt_price_x96 = sqrt(token1/token0) * 2^96
+        let (price_0, price_1) = if zero_for_one {
+            (sell_price_after_fee, buy_price)
+        } else {
+            (buy_price, sell_price_after_fee)
+        };
+
+        // Convert to sqrt price: sqrt(price_1 / price_0) * 2^96
+        get_sqrt_price_q96(price_1, price_0)
+    }
+
+    /// Helper method to execute a swap to a target price limit.
+    /// Returns the swap results containing both amount_in and amount_out.
+    fn execute_swap_to_price_limit(
+        &self,
+        token_in: &Bytes,
+        token_out: &Bytes,
+        target_price: &Price,
+    ) -> Result<Option<(U256, U256, bool)>, SimulationError> {
+        if self.liquidity == 0 {
+            return Ok(None);
+        }
```

Comment:
this should also be in utils and price can potentially be a value not a reference here